### ISSUE TITLE
left hand side vector component selection

### ISSUE THREAD

**Issue by bashbaug** [KhronosGroup/OpenCL-Docs#581]
We were recently surprised internally by "left-hand-side" vector component selection issue, and we also had a related CTS issue (https://github.com/KhronosGroup/OpenCL-CTS/issues/1120), so I think this is an area we need to clarify in the spec:

The question has to do with the meaning of left-hand-side component selection when applied to a vector in memory.  Specifically, what should happen when a single vector component in memory is selected and written to?

See example: https://godbolt.org/z/Kjbhcc

```c
dst[0].z = 5;
```

One interpretation - that appears to be especially common among graphics people - is that this is a single-component write to memory.

Clang has a different interpretation, and seems to be treating this as a vector read, followed by a vector modification (insertion), followed by a vector write:

```llvm
  %2 = load <4 x i32>, <4 x i32> addrspace(1)* %0, align 16       ;; READ
  %3 = insertelement <4 x i32> %2, i32 5, i64 2                   ;; MODIFY
  store <4 x i32> %3, <4 x i32> addrspace(1)* %0, align 16        ;; WRITE
```

In most cases, the two different interpretations are functionally identical (there certainly could be a performance difference!), but in other cases the second interpretation can lead to unintended race conditions:

```c
// Imagine that one work-item taking the "if" branch executes concurrently
// with another work-item taking the "else" branch.  If these are read-
// modify-write operations, this is a race condition:
if (some_condition) {
    dst[0].x = 5;
} else {
    dst[0].y = 6;
}
```

Which interpretation is correct?  Is the code above a race condition?  Can this be described more clearly in the spec?  Thanks!

**Comment by Oblomov** [comment]
I suspect this would have to be _at best_ implementation- and vector-width dependent. Consider for example an OpenCL 1.0 platform without `cl_khr_byte_addressable_store` support, and `dst`is a `char4` rather tha an `int4`. Then it would be impossible for the platform to write to `dst[0].x` directly, and the instructions would inevitably map to a load+store of the entire referenced `char4` anyway.

One could also argue that the spec text “it is an error to take the address of a vector element” is an argument in favor of clang's interpretation (you cannot freely address individual vector components), even though the examples specifically refer to a _private_ vector.

Updates to individual components (or groups thereof) can still be obtained using `vstore` or pointer casting. E.g. in reference to the example:

    global int * d = dst + 0; // pointer to 0-th element in dst
    if (some_condition) { d[0] = 5; /* update .x */ } else { d[1] = 6; /* update .y */ }

and this will result in individual component updates, as expected, provided the hardware supports them (e.g. if using `char` instead of `int` it would fail to compile on a 1.0 platform without `cl_khr_byte_addressable_store`).

User-wise, I think arguments can be made for both sides: being able to write to individual vector components without pointer shuffling is definitely nicer, but at the same time if you are indeed working with vector types it does make sense that they are considered “native” enough to warrant treatment as a unit.

**Comment by AnastasiaStulova** [comment]
I would agree that Clang and LLVM view vectors quite differently to arrays. While arrays are a collection of elements of a certain type located consequently in memory, vectors are undividable types that have components. Therefore the sizes of vectors are predefined and can't be arbitrary like for arrays.

It seems like we should at least adequately reflect this in the documentation to allow write parallel code safely.
